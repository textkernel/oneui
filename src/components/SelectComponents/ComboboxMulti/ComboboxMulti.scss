.ComboboxMulti {
    &__wrapper {
        cursor: pointer;
        display: flex;
        flex-wrap: wrap;
        align-items: center;
        height: 34px;
        box-sizing: border-box;
        outline: none;
        padding: var(--spacing-normal);
        padding-bottom: var(--margin);

        &--focused {
            height: auto;
        }

        &--disabled {
            cursor: not-allowed;
            background-color: var(--color-neutral-25);
        }
    }

    &__input {
        border: 0;
        margin: var(--spacing-normal);
        padding-left: var(--spacing-normal);
        padding-right: var(--spacing-normal);
        flex: 1;
        font-size: var(--font-size-normal);
        background-color: var(--color-background);
        min-width: 150px;

        &:focus {
            outline: none;
        }

        &::placeholder {
            color: var(--color-neutral-25);
        }

        &--disabled {
            cursor: not-allowed;
            background-color: var(--color-neutral-25);
        }
    }

    &__dropdownIcon {
        display: inline-block;
        vertical-align: middle;
        font-size: 20px;
        margin-right: var(--spacing-normal);
        color: var(--color-neutral-60);

        &--disabled {
            color: var(--color-neutral-75);
        }
    }
}
