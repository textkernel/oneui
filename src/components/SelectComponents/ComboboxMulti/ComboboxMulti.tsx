import * as React from 'react';
import { IoMdArrowDropup, IoMdArrowDropdown } from 'react-icons/io';
import { bem } from '../../../utils/bem/bem';
import { SelectBase, BasicSelectProps, SelectInputFieldProps } from '../SelectBase';
import { SuggestionsList } from '../SuggestionsList';
import styles from './ComboboxMulti.scss';
import { ESCAPE_KEY } from '../../../constants';

const { elem } = bem('ComboboxMulti', styles);

export interface Props<S> extends BasicSelectProps<S>, SelectInputFieldProps {
    /** define id for input element */
    id?: string;
    /** to be shown in the input field when no value is typed */
    inputPlaceholder: string;
    /** to be shown when no suggestions are available */
    noSuggestionsPlaceholder: string;
    /** enable ListOptimizer component for decreasing render time */
    useOptimizeListRender?: boolean;
    /** Additional HTML attributes to be applied to the input element */
    inputAttrs?: DictionaryOf<string | boolean>;
}

/**
 * ## Usage information
 * This component is recommended to use when there's a static known list of values.
 * The user can filter the list through the input field. The list is shown right away
 * by clicking on the control. Selected options are not rendered by the component itself.
 * The application using this component should rendered them separately,
 * e.g. above this component using SelectedOption.
 *
 * More detailed face-to-face comparison of Select components can be found
 * [here](https://docs.google.com/spreadsheets/d/1VyYR54RpNaPWLBXOoBPkFEkmzLS_LfEEGdm1ZTTOcHU/edit#gid=0)
 */
export function ComboboxMulti<S>({
    id = undefined,
    onSelectionAdd,
    inputRef: inputRefFromProps,
    suggestions,
    suggestionToString,
    suggestionItemRenderer,
    noSuggestionsPlaceholder,
    onBlur,
    onInputValueChange,
    disabled,
    inputPlaceholder,
    useOptimizeListRender = false,
    inputAttrs = {},
    ...rest
}: Props<S>) {
    const inputRef = inputRefFromProps || React.createRef<HTMLInputElement>();

    // eslint-disable-next-line react/display-name
    const renderFocused = ({ getInputProps, getToggleButtonProps, onBlur: blur }) => {
        const handleInputKeyDown = (event) => {
            if (event.key === ESCAPE_KEY) {
                inputRef.current?.blur();
            }
        };

        return (
            <div tabIndex={0} role="searchbox" {...elem('wrapper', { disabled })}>
                <input
                    {...getInputProps({
                        ...inputAttrs,
                        id,
                        disabled,
                        ref: inputRef,
                        placeholder: inputPlaceholder,
                        onKeyDown: handleInputKeyDown,
                        'data-lpignore': true,
                        ...elem('input', { disabled }),
                    })}
                />
                <IoMdArrowDropup
                    {...elem('dropdownIcon', { disabled })}
                    {...getToggleButtonProps({ disabled, onClick: blur })}
                />
            </div>
        );
    };

    // eslint-disable-next-line react/display-name
    const renderBlurred = ({ getInputProps, getToggleButtonProps, onFocus }) => (
        <div tabIndex={0} role="searchbox" {...elem('wrapper', { disabled })}>
            <input
                {...getInputProps({
                    ...inputAttrs,
                    id,
                    disabled,
                    ref: inputRef,
                    placeholder: inputPlaceholder,
                    onFocus,
                    'data-lpignore': true,
                    ...elem('input', { disabled }),
                })}
            />
            <IoMdArrowDropdown
                {...elem('dropdownIcon', { disabled })}
                {...getToggleButtonProps({
                    disabled,
                    onClick: (e) => {
                        e?.stopPropagation();
                    },
                })}
            />
        </div>
    );

    return (
        <SelectBase
            {...rest}
            suggestions={suggestions}
            suggestionToString={suggestionToString}
            disabled={disabled}
            inputRef={inputRef}
            onBlur={onBlur}
            onSelectionAdd={onSelectionAdd}
            onInputValueChange={onInputValueChange}
            listRenderer={(listProps) => (
                <SuggestionsList
                    {...listProps}
                    isLoading={false}
                    useOptimizeRender={useOptimizeListRender}
                    noSuggestionsPlaceholder={noSuggestionsPlaceholder}
                    suggestionItemRenderer={suggestionItemRenderer}
                    passDisabledToListItems
                />
            )}
            focusedRenderer={renderFocused}
            blurredRenderer={renderBlurred}
            keepExpandedAfterSelection
        />
    );
}

ComboboxMulti.displayName = 'ComboboxMulti';
