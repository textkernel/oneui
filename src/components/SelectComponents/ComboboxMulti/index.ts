export { ComboboxMulti, Props as ComboboxMultiProps } from './ComboboxMulti';
