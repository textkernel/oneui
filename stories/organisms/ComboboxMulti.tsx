import * as React from 'react';
import { ComboboxMulti } from '@textkernel/oneui';
import { SUGGESTIONS, SUGGESTION_TO_STRING } from '../../src/constants/suggestions';

type TSuggestion = { name: string };

export default {
    title: 'Organisms/Select Components/ComboboxMulti',
    component: ComboboxMulti,
};

export const _ComboboxMulti = (args) => {
    const [selectedSuggestions, setSelectedSuggestions] = React.useState<TSuggestion[]>([]);
    const [inputValue, setInputValue] = React.useState('');
    const [disabled, setDisabled] = React.useState(false);

    const getSuggestions = (): TSuggestion[] => {
        return SUGGESTIONS.filter(
            (item) =>
                !selectedSuggestions.includes(item) &&
                item.name.toLocaleLowerCase().includes(inputValue.toLocaleLowerCase())
        );
    };

    const onInputValueChange = (value: string) => {
        console.log(`onInputValueChange was called with ${value}`);
        setInputValue(value);
    };

    const onSelectionAdd = (item: TSuggestion) => {
        console.log(`onSelectionAdd was called with {name: ${item?.name}}`);
        setSelectedSuggestions([...selectedSuggestions, item]);
        if (selectedSuggestions.length === SUGGESTIONS.length) {
            setDisabled(true);
        }
    };

    const onFocus = () => {
        console.log('onFocus was called');
    };

    const onBlur = () => {
        console.log('onBlur was called');
        setTimeout(() => setInputValue(''), 200);
    };

    return (
        <>
            <ul>
                {selectedSuggestions.map((item) => (
                    <li key={item.name}>{item.name}</li>
                ))}
            </ul>
            <ComboboxMulti<TSuggestion>
                {...args}
                suggestions={getSuggestions()}
                onFocus={onFocus}
                onBlur={onBlur}
                onSelectionAdd={onSelectionAdd}
                onInputValueChange={onInputValueChange}
                disabled={args.disabled || disabled}
            />
        </>
    );
};
_ComboboxMulti.args = {
    style: { width: '650px' },
    useOptimizeListRender: false,
    inputPlaceholder: 'Select something...',
    noSuggestionsPlaceholder: 'No suggestions found...',
    suggestionToString: SUGGESTION_TO_STRING,
};
