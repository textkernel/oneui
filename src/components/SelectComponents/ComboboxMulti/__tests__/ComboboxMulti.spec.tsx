import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import { ComboboxMulti } from '../ComboboxMulti';
import {
    SUGGESTIONS,
    SUGGESTION_TO_STRING,
} from '../../../AutosuggestDeprecated/__mocks__/suggestions';

describe('ComboboxMulti', () => {
    const suggestionToString = SUGGESTION_TO_STRING;
    const inputPlaceholder = 'type here...';
    const noSuggestionsPlaceholder = 'No suggestions...';
    const upArrowLabel = 'up arrow';
    const downArrowLabel = 'down arrow';
    const mockOnSelectionAdd = jest.fn();
    const mockOnInputValueChange = jest.fn();
    const mockOnBlur = jest.fn();

    const suggestions = SUGGESTIONS;

    let view;
    let inputNodeField;

    const setFocus = async (user) => user.click(screen.getByRole('listbox'));

    const defaultProps = {
        suggestions,
        suggestionToString,
        inputPlaceholder,
        noSuggestionsPlaceholder,
        downArrowLabel,
        upArrowLabel,
        onSelectionAdd: mockOnSelectionAdd,
        onInputValueChange: mockOnInputValueChange,
        onBlur: mockOnBlur,
    };
    const rerenderView = (props) => {
        view.rerender(<ComboboxMulti {...defaultProps} {...props} />);
    };

    beforeEach(() => {
        view = render(<ComboboxMulti {...defaultProps} />);
        inputNodeField = screen.getByRole('textbox');
    });

    describe('rendering', () => {
        it('should initially render empty component correctly', () => {
            expect(view.container).toMatchSnapshot();
            expect(screen.getByLabelText(downArrowLabel)).toBeInTheDocument();
        });

        it('should add additional attributes to input field when component is blurred', () => {
            const newProps = {
                inputAttrs: { 'data-test': true, title: 'some title' },
            };
            rerenderView(newProps);

            expect(view.container).toMatchSnapshot();
            const inputField = screen.getAllByRole('textbox')[0];

            expect(inputField).toHaveAttribute('data-test', 'true');
            expect(inputField).toHaveAttribute('title', 'some title');
        });

        it('should set focus on the input field', async () => {
            const user = userEvent.setup();
            await setFocus(user);

            expect(inputNodeField).toBe(document.activeElement);
        });

        it('should not set focus on the input field when the component is disabled', async () => {
            const user = userEvent.setup();
            const newProps = {
                disabled: true,
            };
            rerenderView(newProps);
            await setFocus(user);

            expect(inputNodeField).not.toBe(document.activeElement);
        });

        it('should add additional attributes to input field when component is focused', async () => {
            const user = userEvent.setup();
            const newProps = {
                inputAttrs: { 'data-test': true, title: 'some title' },
            };
            rerenderView(newProps);
            await setFocus(user);
            const inputField = screen.getAllByRole('textbox')[0];

            expect(inputField).toHaveAttribute('data-test', 'true');
            expect(inputField).toHaveAttribute('title', 'some title');
        });

        it('should render noSuggestions placeholder when empty suggestions list is passed', async () => {
            const user = userEvent.setup();
            const newProps = {
                suggestions: [],
            };
            rerenderView(newProps);
            await setFocus(user);

            expect(view.container).toMatchSnapshot();
            expect(screen.queryByRole('option')).not.toBeInTheDocument();
            expect(screen.getByRole('listitem')).toBeInTheDocument();
            expect(screen.getByText(noSuggestionsPlaceholder)).toBeInTheDocument();
        });

        it('should render selection placeholder when component is not focused', () => {
            const inputField = screen.getAllByRole('textbox')[0];

            expect(screen.queryByPlaceholderText(inputPlaceholder)).toBeInTheDocument();
            expect(inputField.getAttribute('placeholder')).toEqual(inputPlaceholder);
        });
    });
});
