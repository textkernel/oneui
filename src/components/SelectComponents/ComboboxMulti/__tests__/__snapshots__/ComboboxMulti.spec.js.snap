// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComboboxMulti rendering should initially render empty component correctly 1`] = `
<ComboboxMulti
  inputPlaceholder="type here..."
  noSuggestionsPlaceholder="No suggestions..."
  onBlur={[MockFunction]}
  onInputValueChange={[MockFunction]}
  onSelectionAdd={[MockFunction]}
  suggestionToString={[Function]}
  suggestions={
    Array [
      Object {
        "name": "Sun",
      },
      Object {
        "name": "Moon",
      },
      Object {
        "name": "Mercury",
      },
      Object {
        "name": "Venus",
      },
      Object {
        "name": "Earth",
      },
      Object {
        "name": "Mars",
      },
      Object {
        "name": "Jupiter",
      },
      Object {
        "name": "Saturn",
      },
      Object {
        "name": "Neptune",
      },
      Object {
        "disabled": true,
        "name": "Pluto",
      },
    ]
  }
>
  <SelectBase
    blurredRenderer={[Function]}
    focusedRenderer={[Function]}
    inputRef={
      Object {
        "current": <input
          aria-autocomplete="list"
          aria-labelledby="downshift-0-label"
          autocomplete="off"
          class="ComboboxMulti__input"
          data-lpignore="true"
          placeholder="type here..."
          value=""
        />,
      }
    }
    keepExpandedAfterSelection={true}
    listRenderer={[Function]}
    onBlur={[MockFunction]}
    onInputValueChange={[MockFunction]}
    onSelectionAdd={[MockFunction]}
    suggestionToString={[Function]}
    suggestions={
      Array [
        Object {
          "name": "Sun",
        },
        Object {
          "name": "Moon",
        },
        Object {
          "name": "Mercury",
        },
        Object {
          "name": "Venus",
        },
        Object {
          "name": "Earth",
        },
        Object {
          "name": "Mars",
        },
        Object {
          "name": "Jupiter",
        },
        Object {
          "name": "Saturn",
        },
        Object {
          "name": "Neptune",
        },
        Object {
          "disabled": true,
          "name": "Pluto",
        },
      ]
    }
  >
    <div
      className="SelectBase"
    >
      <Downshift
        defaultHighlightedIndex={0}
        defaultIsOpen={false}
        environment={[Window]}
        getA11yStatusMessage={[Function]}
        inputValue=""
        itemToString={[Function]}
        onChange={[Function]}
        onInputValueChange={[Function]}
        onOuterClick={[Function]}
        onSelect={[Function]}
        onStateChange={[Function]}
        onUserAction={[Function]}
        scrollIntoView={[Function]}
        selectedItemChanged={[Function]}
        stateReducer={[Function]}
        suppressRefError={false}
      >
        <div
          aria-expanded={false}
          aria-haspopup="listbox"
          aria-labelledby="downshift-0-label"
          aria-owns={null}
          className="SelectBase__main"
          role="combobox"
        >
          <FieldWrapper
            className="SelectBase__field"
            clearLabel=""
            isArrowUp={false}
            isFocused={false}
            onArrowClick={[Function]}
            onClear={[Function]}
            onClick={[Function]}
            showArrow={false}
            showClearButton={false}
          >
            <div
              className="FieldWrapper SelectBase__field"
              onClick={[Function]}
            >
              <div
                className="FieldWrapper__content"
              >
                <div
                  className="ComboboxMulti__wrapper"
                  role="searchbox"
                  tabIndex={0}
                >
                  <input
                    aria-activedescendant={null}
                    aria-autocomplete="list"
                    aria-controls={null}
                    aria-labelledby="downshift-0-label"
                    autoComplete="off"
                    className="ComboboxMulti__input"
                    data-lpignore={true}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    placeholder="type here..."
                    value=""
                  />
                  <IoMdArrowDropdown
                    aria-haspopup={true}
                    aria-label="open menu"
                    className="ComboboxMulti__dropdownIcon"
                    data-toggle={true}
                    onBlur={[Function]}
                    onClick={[Function]}
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                    role="button"
                    type="button"
                  >
                    <IconBase
                      aria-haspopup={true}
                      aria-label="open menu"
                      attr={
                        Object {
                          "viewBox": "0 0 512 512",
                        }
                      }
                      className="ComboboxMulti__dropdownIcon"
                      data-toggle={true}
                      onBlur={[Function]}
                      onClick={[Function]}
                      onKeyDown={[Function]}
                      onKeyUp={[Function]}
                      role="button"
                      type="button"
                    >
                      <svg
                        aria-haspopup={true}
                        aria-label="open menu"
                        className="ComboboxMulti__dropdownIcon"
                        data-toggle={true}
                        fill="currentColor"
                        height="1em"
                        onBlur={[Function]}
                        onClick={[Function]}
                        onKeyDown={[Function]}
                        onKeyUp={[Function]}
                        role="button"
                        stroke="currentColor"
                        strokeWidth="0"
                        style={
                          Object {
                            "color": undefined,
                          }
                        }
                        type="button"
                        viewBox="0 0 512 512"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M128 192l128 128 128-128z"
                          key="0"
                        />
                      </svg>
                    </IconBase>
                  </IoMdArrowDropdown>
                </div>
                <List
                  aria-labelledby="downshift-0-label"
                  id="downshift-0-menu"
                  isControlledNavigation={true}
                  role="listbox"
                >
                  <ul
                    aria-labelledby="downshift-0-label"
                    className="List"
                    id="downshift-0-menu"
                    role="listbox"
                  />
                </List>
              </div>
            </div>
          </FieldWrapper>
        </div>
      </Downshift>
    </div>
  </SelectBase>
</ComboboxMulti>
`;

exports[`ComboboxMulti rendering should render all suggestions from the list 1`] = `
<ComboboxMulti
  inputPlaceholder="type here..."
  noSuggestionsPlaceholder="No suggestions..."
  onBlur={[MockFunction]}
  onInputValueChange={[MockFunction]}
  onSelectionAdd={[MockFunction]}
  suggestionToString={[Function]}
  suggestions={
    Array [
      Object {
        "name": "Sun",
      },
      Object {
        "name": "Moon",
      },
      Object {
        "name": "Mercury",
      },
      Object {
        "name": "Venus",
      },
      Object {
        "name": "Earth",
      },
      Object {
        "name": "Mars",
      },
      Object {
        "name": "Jupiter",
      },
      Object {
        "name": "Saturn",
      },
      Object {
        "name": "Neptune",
      },
      Object {
        "disabled": true,
        "name": "Pluto",
      },
    ]
  }
>
  <SelectBase
    blurredRenderer={[Function]}
    focusedRenderer={[Function]}
    inputRef={
      Object {
        "current": <input
          aria-activedescendant="downshift-6-item-0"
          aria-autocomplete="list"
          aria-controls="downshift-6-menu"
          aria-labelledby="downshift-6-label"
          autocomplete="off"
          class="ComboboxMulti__input"
          data-lpignore="true"
          placeholder="type here..."
          value=""
        />,
      }
    }
    keepExpandedAfterSelection={true}
    listRenderer={[Function]}
    onBlur={[MockFunction]}
    onInputValueChange={[MockFunction]}
    onSelectionAdd={[MockFunction]}
    suggestionToString={[Function]}
    suggestions={
      Array [
        Object {
          "name": "Sun",
        },
        Object {
          "name": "Moon",
        },
        Object {
          "name": "Mercury",
        },
        Object {
          "name": "Venus",
        },
        Object {
          "name": "Earth",
        },
        Object {
          "name": "Mars",
        },
        Object {
          "name": "Jupiter",
        },
        Object {
          "name": "Saturn",
        },
        Object {
          "name": "Neptune",
        },
        Object {
          "disabled": true,
          "name": "Pluto",
        },
      ]
    }
  >
    <div
      className="SelectBase"
    >
      <Downshift
        defaultHighlightedIndex={0}
        defaultIsOpen={false}
        environment={[Window]}
        getA11yStatusMessage={[Function]}
        inputValue=""
        itemToString={[Function]}
        onChange={[Function]}
        onInputValueChange={[Function]}
        onOuterClick={[Function]}
        onSelect={[Function]}
        onStateChange={[Function]}
        onUserAction={[Function]}
        scrollIntoView={[Function]}
        selectedItemChanged={[Function]}
        stateReducer={[Function]}
        suppressRefError={false}
      >
        <div
          aria-expanded={true}
          aria-haspopup="listbox"
          aria-labelledby="downshift-6-label"
          aria-owns="downshift-6-menu"
          className="SelectBase__main SelectBase__main--focused"
          role="combobox"
        >
          <FieldWrapper
            className="SelectBase__field"
            clearLabel=""
            isArrowUp={true}
            isFocused={true}
            onArrowClick={[Function]}
            onClear={[Function]}
            onClick={[Function]}
            showArrow={false}
            showClearButton={false}
          >
            <div
              className="FieldWrapper FieldWrapper--isFocused SelectBase__field"
              onClick={[Function]}
            >
              <div
                className="FieldWrapper__content"
              >
                <div
                  className="ComboboxMulti__wrapper"
                  role="searchbox"
                  tabIndex={0}
                >
                  <input
                    aria-activedescendant="downshift-6-item-0"
                    aria-autocomplete="list"
                    aria-controls="downshift-6-menu"
                    aria-labelledby="downshift-6-label"
                    autoComplete="off"
                    className="ComboboxMulti__input"
                    data-lpignore={true}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onKeyDown={[Function]}
                    placeholder="type here..."
                    value=""
                  />
                  <IoMdArrowDropup
                    aria-haspopup={true}
                    aria-label="close menu"
                    className="ComboboxMulti__dropdownIcon"
                    data-toggle={true}
                    onBlur={[Function]}
                    onClick={[Function]}
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                    role="button"
                    type="button"
                  >
                    <IconBase
                      aria-haspopup={true}
                      aria-label="close menu"
                      attr={
                        Object {
                          "viewBox": "0 0 512 512",
                        }
                      }
                      className="ComboboxMulti__dropdownIcon"
                      data-toggle={true}
                      onBlur={[Function]}
                      onClick={[Function]}
                      onKeyDown={[Function]}
                      onKeyUp={[Function]}
                      role="button"
                      type="button"
                    >
                      <svg
                        aria-haspopup={true}
                        aria-label="close menu"
                        className="ComboboxMulti__dropdownIcon"
                        data-toggle={true}
                        fill="currentColor"
                        height="1em"
                        onBlur={[Function]}
                        onClick={[Function]}
                        onKeyDown={[Function]}
                        onKeyUp={[Function]}
                        role="button"
                        stroke="currentColor"
                        strokeWidth="0"
                        style={
                          Object {
                            "color": undefined,
                          }
                        }
                        type="button"
                        viewBox="0 0 512 512"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M128 320l128-128 128 128z"
                          key="0"
                        />
                      </svg>
                    </IconBase>
                  </IoMdArrowDropup>
                </div>
                <List
                  aria-labelledby="downshift-6-label"
                  className="SelectBase__list--focused"
                  id="downshift-6-menu"
                  isControlledNavigation={true}
                  role="listbox"
                >
                  <ul
                    aria-labelledby="downshift-6-label"
                    className="List SelectBase__list--focused"
                    id="downshift-6-menu"
                    role="listbox"
                  >
                    <SuggestionsList
                      className="List__item"
                      getItemProps={[Function]}
                      highlightedIndex={0}
                      inputValue=""
                      isLoading={false}
                      key=".0"
                      noSuggestionsPlaceholder="No suggestions..."
                      passDisabledToListItems={true}
                      suggestionToString={[Function]}
                      suggestions={
                        Array [
                          Object {
                            "name": "Sun",
                          },
                          Object {
                            "name": "Moon",
                          },
                          Object {
                            "name": "Mercury",
                          },
                          Object {
                            "name": "Venus",
                          },
                          Object {
                            "name": "Earth",
                          },
                          Object {
                            "name": "Mars",
                          },
                          Object {
                            "name": "Jupiter",
                          },
                          Object {
                            "name": "Saturn",
                          },
                          Object {
                            "name": "Neptune",
                          },
                          Object {
                            "disabled": true,
                            "name": "Pluto",
                          },
                        ]
                      }
                      useOptimizeRender={false}
                    >
                      <ListItem
                        aria-selected={true}
                        disabled={false}
                        id="downshift-6-item-0"
                        isHighlighted={true}
                        key="Sun"
                        onClick={[Function]}
                        onMouseDown={[Function]}
                        onMouseMove={[Function]}
                        passDisabledToLi={true}
                        role="option"
                        style={Object {}}
                      >
                        <li
                          aria-selected={true}
                          className="ListItem ListItem--isHighlighted ListItem--clickable"
                          disabled={false}
                          id="downshift-6-item-0"
                          onClick={[Function]}
                          onMouseDown={[Function]}
                          onMouseMove={[Function]}
                          role="presentation"
                          style={Object {}}
                        >
                          <MarkedText
                            inline={true}
                            key=".0"
                            marker=""
                          >
                            <Text
                              inline={true}
                            >
                              <span>
                                Sun
                              </span>
                            </Text>
                          </MarkedText>
                        </li>
                      </ListItem>
                      <ListItem
                        aria-selected={false}
                        disabled={false}
                        id="downshift-6-item-1"
                        isHighlighted={false}
                        key="Moon"
                        onClick={[Function]}
                        onMouseDown={[Function]}
                        onMouseMove={[Function]}
                        passDisabledToLi={true}
                        role="option"
                        style={Object {}}
                      >
                        <li
                          aria-selected={false}
                          className="ListItem ListItem--clickable"
                          disabled={false}
                          id="downshift-6-item-1"
                          onClick={[Function]}
                          onMouseDown={[Function]}
                          onMouseMove={[Function]}
                          role="presentation"
                          style={Object {}}
                        >
                          <MarkedText
                            inline={true}
                            key=".0"
                            marker=""
                          >
                            <Text
                              inline={true}
                            >
                              <span>
                                Moon
                              </span>
                            </Text>
                          </MarkedText>
                        </li>
                      </ListItem>
                      <ListItem
                        aria-selected={false}
                        disabled={false}
                        id="downshift-6-item-2"
                        isHighlighted={false}
                        key="Mercury"
                        onClick={[Function]}
                        onMouseDown={[Function]}
                        onMouseMove={[Function]}
                        passDisabledToLi={true}
                        role="option"
                        style={Object {}}
                      >
                        <li
                          aria-selected={false}
                          className="ListItem ListItem--clickable"
                          disabled={false}
                          id="downshift-6-item-2"
                          onClick={[Function]}
                          onMouseDown={[Function]}
                          onMouseMove={[Function]}
                          role="presentation"
                          style={Object {}}
                        >
                          <MarkedText
                            inline={true}
                            key=".0"
                            marker=""
                          >
                            <Text
                              inline={true}
                            >
                              <span>
                                Mercury
                              </span>
                            </Text>
                          </MarkedText>
                        </li>
                      </ListItem>
                      <ListItem
                        aria-selected={false}
                        disabled={false}
                        id="downshift-6-item-3"
                        isHighlighted={false}
                        key="Venus"
                        onClick={[Function]}
                        onMouseDown={[Function]}
                        onMouseMove={[Function]}
                        passDisabledToLi={true}
                        role="option"
                        style={Object {}}
                      >
                        <li
                          aria-selected={false}
                          className="ListItem ListItem--clickable"
                          disabled={false}
                          id="downshift-6-item-3"
                          onClick={[Function]}
                          onMouseDown={[Function]}
                          onMouseMove={[Function]}
                          role="presentation"
                          style={Object {}}
                        >
                          <MarkedText
                            inline={true}
                            key=".0"
                            marker=""
                          >
                            <Text
                              inline={true}
                            >
                              <span>
                                Venus
                              </span>
                            </Text>
                          </MarkedText>
                        </li>
                      </ListItem>
                      <ListItem
                        aria-selected={false}
                        disabled={false}
                        id="downshift-6-item-4"
                        isHighlighted={false}
                        key="Earth"
                        onClick={[Function]}
                        onMouseDown={[Function]}
                        onMouseMove={[Function]}
                        passDisabledToLi={true}
                        role="option"
                        style={Object {}}
                      >
                        <li
                          aria-selected={false}
                          className="ListItem ListItem--clickable"
                          disabled={false}
                          id="downshift-6-item-4"
                          onClick={[Function]}
                          onMouseDown={[Function]}
                          onMouseMove={[Function]}
                          role="presentation"
                          style={Object {}}
                        >
                          <MarkedText
                            inline={true}
                            key=".0"
                            marker=""
                          >
                            <Text
                              inline={true}
                            >
                              <span>
                                Earth
                              </span>
                            </Text>
                          </MarkedText>
                        </li>
                      </ListItem>
                      <ListItem
                        aria-selected={false}
                        disabled={false}
                        id="downshift-6-item-5"
                        isHighlighted={false}
                        key="Mars"
                        onClick={[Function]}
                        onMouseDown={[Function]}
                        onMouseMove={[Function]}
                        passDisabledToLi={true}
                        role="option"
                        style={Object {}}
                      >
                        <li
                          aria-selected={false}
                          className="ListItem ListItem--clickable"
                          disabled={false}
                          id="downshift-6-item-5"
                          onClick={[Function]}
                          onMouseDown={[Function]}
                          onMouseMove={[Function]}
                          role="presentation"
                          style={Object {}}
                        >
                          <MarkedText
                            inline={true}
                            key=".0"
                            marker=""
                          >
                            <Text
                              inline={true}
                            >
                              <span>
                                Mars
                              </span>
                            </Text>
                          </MarkedText>
                        </li>
                      </ListItem>
                      <ListItem
                        aria-selected={false}
                        disabled={false}
                        id="downshift-6-item-6"
                        isHighlighted={false}
                        key="Jupiter"
                        onClick={[Function]}
                        onMouseDown={[Function]}
                        onMouseMove={[Function]}
                        passDisabledToLi={true}
                        role="option"
                        style={Object {}}
                      >
                        <li
                          aria-selected={false}
                          className="ListItem ListItem--clickable"
                          disabled={false}
                          id="downshift-6-item-6"
                          onClick={[Function]}
                          onMouseDown={[Function]}
                          onMouseMove={[Function]}
                          role="presentation"
                          style={Object {}}
                        >
                          <MarkedText
                            inline={true}
                            key=".0"
                            marker=""
                          >
                            <Text
                              inline={true}
                            >
                              <span>
                                Jupiter
                              </span>
                            </Text>
                          </MarkedText>
                        </li>
                      </ListItem>
                      <ListItem
                        aria-selected={false}
                        disabled={false}
                        id="downshift-6-item-7"
                        isHighlighted={false}
                        key="Saturn"
                        onClick={[Function]}
                        onMouseDown={[Function]}
                        onMouseMove={[Function]}
                        passDisabledToLi={true}
                        role="option"
                        style={Object {}}
                      >
                        <li
                          aria-selected={false}
                          className="ListItem ListItem--clickable"
                          disabled={false}
                          id="downshift-6-item-7"
                          onClick={[Function]}
                          onMouseDown={[Function]}
                          onMouseMove={[Function]}
                          role="presentation"
                          style={Object {}}
                        >
                          <MarkedText
                            inline={true}
                            key=".0"
                            marker=""
                          >
                            <Text
                              inline={true}
                            >
                              <span>
                                Saturn
                              </span>
                            </Text>
                          </MarkedText>
                        </li>
                      </ListItem>
                      <ListItem
                        aria-selected={false}
                        disabled={false}
                        id="downshift-6-item-8"
                        isHighlighted={false}
                        key="Neptune"
                        onClick={[Function]}
                        onMouseDown={[Function]}
                        onMouseMove={[Function]}
                        passDisabledToLi={true}
                        role="option"
                        style={Object {}}
                      >
                        <li
                          aria-selected={false}
                          className="ListItem ListItem--clickable"
                          disabled={false}
                          id="downshift-6-item-8"
                          onClick={[Function]}
                          onMouseDown={[Function]}
                          onMouseMove={[Function]}
                          role="presentation"
                          style={Object {}}
                        >
                          <MarkedText
                            inline={true}
                            key=".0"
                            marker=""
                          >
                            <Text
                              inline={true}
                            >
                              <span>
                                Neptune
                              </span>
                            </Text>
                          </MarkedText>
                        </li>
                      </ListItem>
                      <ListItem
                        aria-selected={false}
                        disabled={true}
                        id="downshift-6-item-9"
                        isHighlighted={false}
                        key="Pluto"
                        onMouseDown={[Function]}
                        passDisabledToLi={true}
                        role="option"
                        style={Object {}}
                      >
                        <li
                          aria-selected={false}
                          className="ListItem ListItem--disabled"
                          disabled={true}
                          id="downshift-6-item-9"
                          onMouseDown={[Function]}
                          role="presentation"
                          style={Object {}}
                        >
                          <MarkedText
                            inline={true}
                            key=".0"
                            marker=""
                          >
                            <Text
                              inline={true}
                            >
                              <span>
                                Pluto
                              </span>
                            </Text>
                          </MarkedText>
                        </li>
                      </ListItem>
                    </SuggestionsList>
                  </ul>
                </List>
              </div>
            </div>
          </FieldWrapper>
        </div>
      </Downshift>
    </div>
  </SelectBase>
</ComboboxMulti>
`;
